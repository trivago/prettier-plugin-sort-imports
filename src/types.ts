import { RequiredOptions } from 'prettier';
import { ParserPlugin } from '@babel/parser';

export interface PrettierOptions extends RequiredOptions {
    importOrder: string[];
    importOrderSeparation: boolean;
    experimentalBabelParserPluginsList: ParserPlugin[];
}
