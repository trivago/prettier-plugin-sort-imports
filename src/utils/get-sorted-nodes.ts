import {
    chunkSideEffectNode,
    chunkSideOtherNode,
    newLineNode,
} from '../constants';
import { GetSortedNodes, ImportChunk, ImportOrLine } from '../types';
import { adjustCommentsOnSortedNodes } from './adjust-comments-on-sorted-nodes';
import { getSortedNodesByImportOrder } from './get-sorted-nodes-by-import-order';

/**
 * This function returns the given nodes, sorted in the order as indicated by
 * the importOrder array. The plugin considers these import nodes as local
 * import declarations
 *
 * In addition, this method preserves the relative order of side effect imports
 * and non side effect imports. A side effect import is an ImportDeclaration
 * without any import specifiers. It does this by splitting the import nodes at
 * each side effect node, then sorting only the non side effect import nodes
 * between the side effect nodes according to the given options.
 * @param nodes All import nodes that should be sorted.
 * @param options Options to influence the behavior of the sorting algorithm.
 */
export const getSortedNodes: GetSortedNodes = (nodes, options) => {
    const { importOrderSeparation, importOrderSideEffects } =
        options;

    // Split nodes at each boundary between a side-effect node and a
    // non-side-effect node, keeping both types of nodes together.
    const splitBySideEffectNodes = nodes.reduce<ImportChunk[]>(
        (chunks, node) => {
            const isChunkEffectNode =
                node.specifiers.length === 0 &&
                importOrderSideEffects === false;
            const type = isChunkEffectNode
                ? chunkSideEffectNode
                : chunkSideOtherNode;
            const last = chunks[chunks.length - 1];
            if (last === undefined || last.type !== type) {
                chunks.push({ type, nodes: [node] });
            } else {
                last.nodes.push(node);
            }
            return chunks;
        },
        [],
    );

    const finalNodes: ImportOrLine[] = [];

    // Sort each chunk of side-effect and non-side-effect nodes, and insert new
    // lines according the importOrderSeparation option.
    for (const chunk of splitBySideEffectNodes) {
        if (chunk.type === chunkSideEffectNode) {
            // do not sort side effect nodes
            finalNodes.push(...chunk.nodes);
        } else {
            // sort non-side effect nodes
            const sorted = getSortedNodesByImportOrder(chunk.nodes, options);
            finalNodes.push(...sorted);
        }
        if (importOrderSeparation) {
            finalNodes.push(newLineNode);
        }
    }

    if (finalNodes.length > 0 && !importOrderSeparation) {
        finalNodes.push(newLineNode);
    }

    // Adjust the comments on the sorted nodes to match the original comments
    adjustCommentsOnSortedNodes(nodes, finalNodes);

    return finalNodes;
};
