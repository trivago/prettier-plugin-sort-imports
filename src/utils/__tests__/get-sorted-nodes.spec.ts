import { parse as babelParser, ParserOptions } from '@babel/parser';
import traverse, { NodePath } from '@babel/traverse';
import { ImportDeclaration, isTSModuleDeclaration } from '@babel/types';
import { getSortedNodes } from '../get-sorted-nodes';

const code = `// first comment
// second comment
import z from 'z';
import c from 'c';
import g from 'g';
import t from 't';
import k from 'k';
import a from 'a';
`;

const getImportNodes = (code: string, options?: ParserOptions) => {
    const importNodes: ImportDeclaration[] = [];
    const ast = babelParser(code, {
        ...options,
        sourceType: 'module',
    });

    traverse(ast, {
        ImportDeclaration(path: NodePath<ImportDeclaration>) {
            const tsModuleParent = path.findParent((p) =>
                isTSModuleDeclaration(p),
            );
            if (!tsModuleParent) {
                importNodes.push(path.node);
            }
        },
    });

    return importNodes;
};

const getSortedNodesNames = (imports: ImportDeclaration[]) =>
    imports
        .filter((i) => i.type === 'ImportDeclaration')
        .map((i) => i.source.value); // TODO: get from specifier

test('it returns all sorted nodes', () => {
    const result = getImportNodes(code);
    const sorted = getSortedNodes(result, [], false) as ImportDeclaration[];
    expect(sorted).toMatchSnapshot();
    expect(getSortedNodesNames(sorted)).toEqual(['a', 'c', 'g', 'k', 't', 'z']);
});

test('it returns all sorted nodes with sort order', () => {
    const result = getImportNodes(code);
    const sorted = getSortedNodes(
        result,
        ['^a$', '^t$', '^k$'],
        true,
    ) as ImportDeclaration[];
    expect(sorted).toMatchSnapshot();
    expect(getSortedNodesNames(sorted)).toEqual(['c', 'g', 'z', 'a', 't', 'k']);
});
