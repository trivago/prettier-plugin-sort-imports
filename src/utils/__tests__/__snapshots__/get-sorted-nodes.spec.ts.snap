// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it returns all sorted nodes 1`] = `
Array [
  Node {
    "end": 148,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": Array [
      Object {
        "end": 16,
        "loc": SourceLocation {
          "end": Position {
            "column": 16,
            "line": 1,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "start": 0,
        "type": "CommentLine",
        "value": " first comment",
      },
      Object {
        "end": 34,
        "loc": SourceLocation {
          "end": Position {
            "column": 17,
            "line": 2,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 0,
            "line": 2,
          },
        },
        "start": 17,
        "type": "CommentLine",
        "value": " second comment",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 8,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 8,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 147,
      "extra": Object {
        "raw": "'a'",
        "rawValue": "a",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 8,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 8,
        },
      },
      "range": undefined,
      "start": 144,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "a",
    },
    "specifiers": Array [
      Node {
        "end": 138,
        "extra": undefined,
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 8,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 8,
          },
        },
        "local": Node {
          "end": 138,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 8,
            },
            "filename": undefined,
            "identifierName": "a",
            "start": Position {
              "column": 7,
              "line": 8,
            },
          },
          "name": "a",
          "range": undefined,
          "start": 137,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 137,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 130,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Node {
    "end": 72,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": Array [],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 4,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 4,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 71,
      "extra": Object {
        "raw": "'c'",
        "rawValue": "c",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 4,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 4,
        },
      },
      "range": undefined,
      "start": 68,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "c",
    },
    "specifiers": Array [
      Node {
        "end": 62,
        "extra": undefined,
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 4,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 4,
          },
        },
        "local": Node {
          "end": 62,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 4,
            },
            "filename": undefined,
            "identifierName": "c",
            "start": Position {
              "column": 7,
              "line": 4,
            },
          },
          "name": "c",
          "range": undefined,
          "start": 61,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 61,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 54,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Node {
    "end": 91,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": Array [],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 5,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 5,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 90,
      "extra": Object {
        "raw": "'g'",
        "rawValue": "g",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 5,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 5,
        },
      },
      "range": undefined,
      "start": 87,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "g",
    },
    "specifiers": Array [
      Node {
        "end": 81,
        "extra": undefined,
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 5,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 5,
          },
        },
        "local": Node {
          "end": 81,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 5,
            },
            "filename": undefined,
            "identifierName": "g",
            "start": Position {
              "column": 7,
              "line": 5,
            },
          },
          "name": "g",
          "range": undefined,
          "start": 80,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 80,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 73,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Node {
    "end": 129,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": Array [],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 7,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 7,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 128,
      "extra": Object {
        "raw": "'k'",
        "rawValue": "k",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 7,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 7,
        },
      },
      "range": undefined,
      "start": 125,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "k",
    },
    "specifiers": Array [
      Node {
        "end": 119,
        "extra": undefined,
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 7,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 7,
          },
        },
        "local": Node {
          "end": 119,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 7,
            },
            "filename": undefined,
            "identifierName": "k",
            "start": Position {
              "column": 7,
              "line": 7,
            },
          },
          "name": "k",
          "range": undefined,
          "start": 118,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 118,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 111,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Node {
    "end": 110,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": Array [],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 6,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 6,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 109,
      "extra": Object {
        "raw": "'t'",
        "rawValue": "t",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 6,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 6,
        },
      },
      "range": undefined,
      "start": 106,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "t",
    },
    "specifiers": Array [
      Node {
        "end": 100,
        "extra": undefined,
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 6,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 6,
          },
        },
        "local": Node {
          "end": 100,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 6,
            },
            "filename": undefined,
            "identifierName": "t",
            "start": Position {
              "column": 7,
              "line": 6,
            },
          },
          "name": "t",
          "range": undefined,
          "start": 99,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 99,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 92,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Node {
    "end": 53,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": null,
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 3,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 3,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 52,
      "extra": Object {
        "raw": "'z'",
        "rawValue": "z",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 3,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 3,
        },
      },
      "range": undefined,
      "start": 49,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "z",
    },
    "specifiers": Array [
      Node {
        "end": 43,
        "extra": undefined,
        "innerComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 3,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 3,
          },
        },
        "local": Node {
          "end": 43,
          "extra": undefined,
          "innerComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 3,
            },
            "filename": undefined,
            "identifierName": "z",
            "start": Position {
              "column": 7,
              "line": 3,
            },
          },
          "name": "z",
          "range": undefined,
          "start": 42,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 42,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 35,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Object {
    "expression": Object {
      "type": "StringLiteral",
      "value": "PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE",
    },
    "innerComments": null,
    "leadingComments": Array [],
    "trailingComments": null,
    "type": "ExpressionStatement",
  },
]
`;

exports[`it returns all sorted nodes with sort order 1`] = `
Array [
  Node {
    "end": 72,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": Array [
      Object {
        "end": 16,
        "loc": SourceLocation {
          "end": Position {
            "column": 16,
            "line": 1,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "start": 0,
        "type": "CommentLine",
        "value": " first comment",
      },
      Object {
        "end": 34,
        "loc": SourceLocation {
          "end": Position {
            "column": 17,
            "line": 2,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 0,
            "line": 2,
          },
        },
        "start": 17,
        "type": "CommentLine",
        "value": " second comment",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 4,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 4,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 71,
      "extra": Object {
        "raw": "'c'",
        "rawValue": "c",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 4,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 4,
        },
      },
      "range": undefined,
      "start": 68,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "c",
    },
    "specifiers": Array [
      Node {
        "end": 62,
        "extra": undefined,
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 4,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 4,
          },
        },
        "local": Node {
          "end": 62,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 4,
            },
            "filename": undefined,
            "identifierName": "c",
            "start": Position {
              "column": 7,
              "line": 4,
            },
          },
          "name": "c",
          "range": undefined,
          "start": 61,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 61,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 54,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Node {
    "end": 91,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": Array [],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 5,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 5,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 90,
      "extra": Object {
        "raw": "'g'",
        "rawValue": "g",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 5,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 5,
        },
      },
      "range": undefined,
      "start": 87,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "g",
    },
    "specifiers": Array [
      Node {
        "end": 81,
        "extra": undefined,
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 5,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 5,
          },
        },
        "local": Node {
          "end": 81,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 5,
            },
            "filename": undefined,
            "identifierName": "g",
            "start": Position {
              "column": 7,
              "line": 5,
            },
          },
          "name": "g",
          "range": undefined,
          "start": 80,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 80,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 73,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Node {
    "end": 53,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": null,
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 3,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 3,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 52,
      "extra": Object {
        "raw": "'z'",
        "rawValue": "z",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 3,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 3,
        },
      },
      "range": undefined,
      "start": 49,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "z",
    },
    "specifiers": Array [
      Node {
        "end": 43,
        "extra": undefined,
        "innerComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 3,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 3,
          },
        },
        "local": Node {
          "end": 43,
          "extra": undefined,
          "innerComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 3,
            },
            "filename": undefined,
            "identifierName": "z",
            "start": Position {
              "column": 7,
              "line": 3,
            },
          },
          "name": "z",
          "range": undefined,
          "start": 42,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 42,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 35,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Object {
    "expression": Object {
      "type": "StringLiteral",
      "value": "PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE",
    },
    "innerComments": null,
    "leadingComments": Array [],
    "trailingComments": null,
    "type": "ExpressionStatement",
  },
  Node {
    "end": 148,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": Array [],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 8,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 8,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 147,
      "extra": Object {
        "raw": "'a'",
        "rawValue": "a",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 8,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 8,
        },
      },
      "range": undefined,
      "start": 144,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "a",
    },
    "specifiers": Array [
      Node {
        "end": 138,
        "extra": undefined,
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 8,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 8,
          },
        },
        "local": Node {
          "end": 138,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 8,
            },
            "filename": undefined,
            "identifierName": "a",
            "start": Position {
              "column": 7,
              "line": 8,
            },
          },
          "name": "a",
          "range": undefined,
          "start": 137,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 137,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 130,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Object {
    "expression": Object {
      "type": "StringLiteral",
      "value": "PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE",
    },
    "innerComments": null,
    "leadingComments": Array [],
    "trailingComments": null,
    "type": "ExpressionStatement",
  },
  Node {
    "end": 110,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": Array [],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 6,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 6,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 109,
      "extra": Object {
        "raw": "'t'",
        "rawValue": "t",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 6,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 6,
        },
      },
      "range": undefined,
      "start": 106,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "t",
    },
    "specifiers": Array [
      Node {
        "end": 100,
        "extra": undefined,
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 6,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 6,
          },
        },
        "local": Node {
          "end": 100,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 6,
            },
            "filename": undefined,
            "identifierName": "t",
            "start": Position {
              "column": 7,
              "line": 6,
            },
          },
          "name": "t",
          "range": undefined,
          "start": 99,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 99,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 92,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Object {
    "expression": Object {
      "type": "StringLiteral",
      "value": "PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE",
    },
    "innerComments": null,
    "leadingComments": Array [],
    "trailingComments": null,
    "type": "ExpressionStatement",
  },
  Node {
    "end": 129,
    "extra": undefined,
    "innerComments": null,
    "leadingComments": Array [],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 7,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 7,
      },
    },
    "range": undefined,
    "source": Node {
      "end": 128,
      "extra": Object {
        "raw": "'k'",
        "rawValue": "k",
      },
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 17,
          "line": 7,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 14,
          "line": 7,
        },
      },
      "range": undefined,
      "start": 125,
      "trailingComments": undefined,
      "type": "StringLiteral",
      "value": "k",
    },
    "specifiers": Array [
      Node {
        "end": 119,
        "extra": undefined,
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 7,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 7,
            "line": 7,
          },
        },
        "local": Node {
          "end": 119,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 7,
            },
            "filename": undefined,
            "identifierName": "k",
            "start": Position {
              "column": 7,
              "line": 7,
            },
          },
          "name": "k",
          "range": undefined,
          "start": 118,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 118,
        "trailingComments": undefined,
        "type": "ImportDefaultSpecifier",
      },
    ],
    "start": 111,
    "trailingComments": null,
    "type": "ImportDeclaration",
  },
  Object {
    "expression": Object {
      "type": "StringLiteral",
      "value": "PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE",
    },
    "innerComments": null,
    "leadingComments": Array [],
    "trailingComments": null,
    "type": "ExpressionStatement",
  },
]
`;
